src/cephadm/cephadm

```diff
@@ -5212,6 +5222,185 @@ def command_gather_facts():
 ##################################
 
 
+class CephadmDaemonHandler(BaseHTTPRequestHandler):
+    def do_GET(self):
+        # return the current state of the cache
+        self.send_response(200)
+        self.send_header('Content-type','application/json')
+        self.end_headers()
+        self.wfile.write(bytes(json.dumps(cephadm_cache), 'ascii'))
+    
+    def log_message(self, format, *args):
+        rqst = " ".join(args)
+        logger.info(f"client:{self.address_string()} [{self.log_date_time_string()}] {rqst}")
+
+
+class CephadmHTTPServer(ThreadingMixIn, HTTPServer):
+    allow_reuse_address = True
+    daemon_threads = True
+
+
+def cephadmd_install():
+    # open firewall port
+    # copy self to /usr/bin somewhere
+    # create the systemd unit file
+    # enable the systemd unit file
+    # start the systemd unit
+    pass
+
+def cephadmd_uninstall():
+    # shutdown the systemd unit
+    # disable it and remove it
+    # close the firewall port
+    pass
+
+def cephadmd_upgrade():
+    # copy self to the /usr/bin
+    # restart the systemd unit
+    pass
```

Comment:
As both upgrading *and* installing the daemon should be idempotent, you should simply be able to call `install` here instead.

Reply:
cephadm's daemon_deploy function itself is not idempotent. if the service is up, this function raises an exception since the port requested would be in use - so a simple rerun is not practical. Perhaps if the code only raises if reconfig is false would be a better bet? That way to refresh you could deploy with --reconfig to refresh the daemon code and restart the service

Thoughts?
@sebastian-philipp ^^

Reply:
We had this problem for the other daemons as well. I remember that we had a PR to disable it port check for redeployments and reconfigs. At some place, we have to be idempotent to allow Mgr fail overs. Not being able to be idempotent would make things more complicated elsewhere.

Reply:
+1. That will need to be a subsequent PR though right?